### ISSUE TITLE
Gif animation blocks when I call Glide.xxx.load(gif_url).into(imageView) in onResourceReady() callback

### ISSUE THREAD

**Issue by muzhi1991** [bumptech/glide#913]
In my app,I need to load a image before load a gif. I don't use thumbnail, then meet some problem.
My codes as follows

``` java
holder.img.setLoadImageListener(new MyImageView.BdLoadImageListener() {
                @Override
                public void onSuccess(final Drawable drawable) {
                    holder.gifLoadingView.show();
                    holder.gifPlayView.setVisibility(View.GONE);
                    // load gif. nested call Glide.with().load(gif_url).into(imgview)
                    holder.request =
                                    holder.img.setAsyncImageUrl(mFragment, itemWrapper.entity.imgUrl, imageSize,
                                            drawable, R.drawable.img_error);
                }

                @Override
                public void onFail() {
                }
 });
// load first image
holder.request = holder.img.setAsyncImageUrl(itemWrapper.entity.thumbUrl, imageSize, R.drawable.img_default,
                            R.drawable.img_default);

// custom image view like this
public class MyImageView extends ImageView {
    private static final String TAG = MyImageView.class.getSimpleName();

    private String mUrl = null;
    private ImageLoadListener mImageLoaderListener;
    private BdLoadImageListener mListener;

    // ......some code

    public void setLoadImageListener(BdLoadImageListener listener) {
        mListener = listener;
    }

    public Target setAsyncImageUrl(Fragment fragment, String url, GlideWrapper.ImageSize size, Drawable placeHolder,
            int errorResourceId) {
        mUrl = url;
        ImageLoadListener mImageLoaderListener = new ImageLoadListener(this);
        DrawableTypeRequest<String> load = Glide.with(fragment).load(url);

        DrawableRequestBuilder builder = load.listener(mImageLoaderListener);
        if (placeHolder != null) {
            builder.placeholder(placeHolder);
        }

        if (errorResourceId != -1 && errorResourceId != 0) {
            builder.error(errorResourceId);
        }
        if (size != null) {
            builder.override(size.width, size.height);
        }
        return builder.fitCenter().dontAnimate().into(this);

    }

    static class ImageLoadListener implements RequestListener<String, GlideDrawable> {

        MyImageView mImageView;

        public ImageLoadListener(MyImageView asyncImageView) {
            mImageView = asyncImageView;
        }

        @Override
        public boolean onException(Exception e, String model, Target<GlideDrawable> target, boolean isFirstResource) {
            MyImageView asyncImageView = mImageView;
            if (asyncImageView != null) {
                if (asyncImageView.mListener != null) {
                    asyncImageView.mListener.onFail();
                }
            }
            return false;
        }

        @Override
        public boolean onResourceReady(GlideDrawable resource, String model, Target<GlideDrawable> target,
                boolean isFromMemoryCache, boolean isFirstResource) {
            MyImageView asyncImageView = mImageView;
            if (asyncImageView != null && asyncImageView.mUrl != null) {
                if (asyncImageView.mListener != null) {
                    asyncImageView.mListener.onSuccess(resource);
                }
            }
            return false;
        }

    public interface BdLoadImageListener {
        public void onSuccess(Drawable drawable);

        public void onFail();
    }
}
```

I find that gif just plays first frame then blocks. I use following codes to solve this problem ,but the solution is not cool.
in onSuccess/onResourceReady callback 

``` java
holder.img.post(new Runnable() {
    @Override
    public void run() {
        holder.request =
        holder.img.setAsyncImageUrl(mFragment, itemWrapper.entity.imgUrl, imageSize,
                    drawable, R.drawable.img_error);
    }
});
```

I want to know what causes this problem,so debug source code(glide version 3.6.1). 
In [GifDrawable.java](https://github.com/bumptech/glide/blob/3.0/library/src/main/java/com/bumptech/glide/load/resource/gif/GifDrawable.java) line 262, "getCallback()==null" is true and then stop&&reset.

Could you tell me why this happened? Is there any other solutions?
thank you

**Comment by TWiStErRob** [comment]
You're stealing the drawable from the first load which is unsafe. That drawable you use for the second load's placeholder becomes invalid the moment you call `.into()` because an `ImageView` can only be associated with one Request at a time with the default ImageViewTarget, the exception is the `thumbnail` call where more loads are racing to be shown in on view. That is the way Glide handles your use case. What is the reason for not using it? It looks like you're trying to reinvent the wheel.

**Comment by muzhi1991** [comment]
I use this solution because I need to display first image then display gif rather than download at same time.
if I use Glide.with(this).load(gifUrl).thumbnail(Glide.with(this).load(imgUrl)).into(this). It will download image and gif at same time and I also can't ensure image download successfully.
is there any method to make this safe and easy?

**Comment by TWiStErRob** [comment]
What do you mean by "can't ensure image download successfully"? In both cases you're using Glide's internal loading mechanisms. To get notifications of success/failure on both the gif and normal load you can add listeners to both main and thumbnail requests.

Yes it will start loading at the same time, but by default the priority of the thumbnail is one higher than the main load, so there's something to show to the user quickly.

Is your thumbUrl the GIF? I'm a little confused, because in the title of the issue you say it stops when you start the load. So it must be already displaying, hence thumb (the first load) must be the GIF.

**Comment by muzhi1991** [comment]
Sorry, my English is poor, I try my best to describe it.
1.As the code that I paste above, I start loading gif in the listener of thumb because I want to load them sequentially. not concurrently. I did this just want to control this process and ensure the order. 
> the priority of the thumbnail is one higher than the main load

this sentence dispel my doubts partly :-D

2.the thumb is a jpg file. I have a picture of gif's content, so I don't use load(gifUrl).asBitmap()

3.I may not describe the phenomenon clearly
>  in the title of the issue you say it stops when you start the load

If use my codes, do as follows
* The first time , set thumbUrl into imageview，the thumb download and display, then set gifUrl into imageview, the gif download and play. **Every thing is fine**.
* Next step, I stop gif (by set thumb into imageview again when I click).
* The second time, I repeat first step, the thumb and gif will load from cache orderly, **the problem appears**, thumb image is ok, but gif just play first frame and then stop. So, I think that it may be a bug.(I track code, every thing is fine except that getCallback() returns null in the onFrameReday() of GifDrawable.java!)

note: If I don't steal drawable as placeholder and just request a gif from the listener of thumb , the phenomenon also happens. 

**Is this a code design that users can't request another image in  the listener of a request?** In some situations，I guess that it may be useful

Thank you!

**Comment by TWiStErRob** [comment]
> 1. want to control this process and ensure the order

In this case chaos may be a better choice. You're loading the thumb and then the GIF, always! Even when the GIF is already downloaded and likely in memory cache so it would be ready immediately, you still delay the display by setting the thumb. `.thumbnail()` is designed to have the best UX, **it will never load the GIF and then the thumb**. The thumbnail is only shown if absolutely necessary: that is the GIF load takes longer. Imagine a 10kb high resolution GIF file with extremely good delta encoded frames, but due to complexity the thumbnail is 50k. In this case the main load may complete much faster and show the animation first, which is what you should want.

> 3. may be a bug

I agree now that I see the full flow and code, it should be started automatically. I'll probably try it sometime, but it may be 3 weeks from now, can't promise anything.

> 3. can't request another image in the listener of a request

This is more likely a side effect. The thumb load's request is associated with the `ImageView` via the Target. When you start a new load, this request is retrieved and cleared and the new request is set. BUT since you're starting the load in the `.listener()` that target is still finishing up, and even though it was cleared it is likely still setting the thumb image, whose request was already cleared.

You can always post the next load into the UI thread queue, so there's no collision.

---

Try to return true from onResourceReady when you're starting a new load, i.e. thumb finishes, but false when GIF finishes; let's see what happens. (In general for debugging I think you should try all this logic without your own factory methods and interfaces, just so it's easier to handle and then when it works you can explode it the way you want to.)

In the end I really think you should go with the built-in thumbnail handling. Your code would be much simpler and likely work as you expect.

Tip: you should also add `.diskCacheStrategy(SOURCE)` to the main GIF load, it'll speed up loading by a magnitude. (There are a lot of "slow gif" issues around on GitHub)

**Comment by muzhi1991** [comment]
Thank you for your advices.
Today, I go with the built-in thumbnail handling and meet some problems. I post here, it may be useful for someone.
My app has a listview.  I need to show a list of static pictures which is same as thumbnail. then load and play a gif when user clicks. When user clicks one of item,my codes as follows.
``` java
    Glide.with(fragment).load(gifurl)
    .override(size.width, size.height)
    .diskCacheStrategy(DiskCacheStrategy.SOURCE)
    .thumbnail(
            Glide.with(fragment).load(thumbUrl).override(size.width, size.height)
            .placeholder(R.drawable.img_default)
            .diskCacheStrategy(DiskCacheStrategy.SOURCE)
            .dontAnimate().fitCenter()
    )
    .error(R.drawable.img_error).dontAnimate().fitCenter()
    .into(this);
```
I write these codes s according to giphy demo, it does work, but unfortunately it is not perfect. The **imageView that holds picture blinks** when I use method above.
In my case, "thumbUrl" picture may have been downloaded successfully before user clicks. So, I expect that thumbnail() will reuse these pictures.
I search issues and find #527. the placeholder(imageview.getDrawable())  solution is unsafe like my previous codes. use copy() funciton is useful. other solution seems to be a little complex.

I want to know why not reuse picture, so track codes and find they have different EngineKey. Specifically, they have different Transformation, (GifBitmapWrapperTransformation vs UnitTransformation)
Finally, **both of thumbnail and gif add  dontTransform()**
``` java
    Glide.with(fragment).load(gifurl)
    .override(size.width, size.height)
    .diskCacheStrategy(DiskCacheStrategy.SOURCE)
    .thumbnail(
            Glide.with(fragment).load(thumbUrl).override(size.width, size.height)
            .placeholder(R.drawable.img_default)
            .diskCacheStrategy(DiskCacheStrategy.SOURCE)
            .dontAnimate().dontTransform().fitCenter()
    )
    .error(R.drawable.img_error).dontAnimate().dontTransform().fitCenter()
    .into(this);
```
Now, it seems ok .

**Comment by TWiStErRob** [comment]
There are some redundancies in your last code. Here's my take on it. Feel free to ignore any parts if it doesn't apply for your case. You essentially implemented what proposed at https://github.com/bumptech/glide/issues/527#issuecomment-148840717 without the crossfade.

I assume when you display the initial static image you do this:
```java
Glide
    .with(fragment)
    .load(thumbUrl)
    .override(size.width, size.height) // (2)
    .into(this) // (1) ImageView.scaleType == fitCenter implicit
```

And then on user click:
```java
Glide
    .with(fragment)
    .load(gifurl)
    .dontAnimate()
    .placeholder(R.drawable.img_default) // better to have them in one place
    .error(R.drawable.img_error)
    // this is not necessary, usually Glide can figure out the best size
    //.override(size.width, size.height)
    .diskCacheStrategy(DiskCacheStrategy.SOURCE)
    .thumbnail(
        Glide
            .with(fragment)
            .load(thumbUrl)
            .dontAnimate()
            // the default is RESULT cache and if you use override whereever you load thumbUrl
            // you'll always hit that cache, no need to double-cache it
            // (the above static load already saved it in RESULT cache with the right size)
            //.diskCacheStrategy(DiskCacheStrategy.SOURCE)
            .override(size.width, size.height) // this has to match (2)
            .fitCenter() // this has to match (1)
    )
    .into(this) // ImageView.scaleType == fitCenter implicit
```

`.dontTransform().fitCenter()` is equivalent to just `.fitCenter()`, both override the current transformation (see their codes).

**Comment by muzhi1991** [comment]
you are right, both fitCenter() and dontTransform() set transformation. As far as I know, If you want to hit memory cache ,the following parameters should remain the same when we use default encoder/decoder.
* url 
* size
* transformation,which is related to ImageView's scaleType and Glide's Settings, this is not easy to notice,

I guess that If we just use "source" disk cache strategy with different size/transformation parameters，glide will also hit disk cache, but can't hit memory cache, so imageView will blink. 

Are there other settings need to pay attention if we want to reuse picture(hit memory cache)?
In addition, If we want to hit disk cache,what settings we need to pay attention when using different disk cache strategy?

**Comment by TWiStErRob** [comment]
Source cache is just URL, result cache is the same as memory cache you listed.

Enable [logging](https://github.com/bumptech/glide/wiki/Debugging-and-Error-Handling#debugging-workflow) so you see which key is used when and what happens inside Glide.

For the full size image it is OK just to use the source cache, because it'll likely load in similar time and you can still hit the memory cache. As far as I know thumbnail has a separate cache (default result in my above code) which will hit the memory cache from the grid.

Did you try my code and it was blinking? Double check the XML scaleType and the thumbnail's transformation if they match. Also log the size object used for the grid and the thumbnail, they must be same too.

**Comment by muzhi1991** [comment]
I have tried your code , it works perfectly. Thank you for your reply.

**Comment by TWiStErRob** [comment]
Let's leave this open I want to try to reproduce your original solution and see why it stopped.

**Comment by TWiStErRob** [comment]
I did some testing and I was able to reproduce your case with this code:
(skipped all the sizing and fallback and error drawables to make things simpler)
```java
Glide
	.with(this)
	.load("http://searchengineland.com/figz/wp-content/seloads/2012/04/penguin.jpg")
	.listener(new LoggingListener<>(new RequestListener<String, GlideDrawable>() {
		@Override public boolean onException(Exception e, String model, Target<GlideDrawable> target, boolean isFirstResource) {
			return false;
		}
		@Override public boolean onResourceReady(GlideDrawable resource, String model, Target<GlideDrawable> target, boolean isFromMemoryCache, boolean isFirstResource) {
			Glide
				.with(TestFragment.this)
				.load("http://bestanimations.com/Animals/Birds/Penguins/animated-penguin-gif-5.gif")
				.diskCacheStrategy(DiskCacheStrategy.SOURCE)
				.listener(new LoggingListener<String, GlideDrawable>())
				.into(imageView);
			return false;
		}
	}))
	.into(imageView);
```

I didn't remember why it's unsafe to call `into()` from `onResourceReady`, now I do.

Consider the first load when the memory cache doesn't have any resources in it. JPG is the thumbnail and GIF is the full image.
The flow goes like this:
 1. JPG load queued
 1. JPG load started
 1. JPG decoded from disk cache
 1. JPG onResourceReady
 1. GIF load queued
 1. JPG set into view
 1. GIF load started
 1. GIF decoded from disk cache
 1. GIF set into view

Now let's assume that the GIF resource is in memory cache.
The JPG resource doesn't matter in this case because if it doesn't hit the cache everything below will be just delayed a little and 1-3 could be replaced by "JPG hits memory cache".
 1. JPG load queued
 1. JPG load started
 1. JPG decoded from disk cache
 1. JPG onResourceReady
 1. GIF hits memory cache
 1. GIF set into view
 1. JPG set into view

The problem here is that if the memory cache is hit, then `into` completes synchronously which in turn implies that `onResourceReady` completes synchronously. So even though the inner load cancels the outer one and then completes sync, the outer one still finishes, because it was about one line away from finishing. Your `post` workaround worked because it lets the JPG complete and only after that it starts the GIF load. So you were likely seeing the thumbnail, and not the first frame of the GIF, though from the logs I see that the first frame of the GIF is loaded. (@sjudd it could be possible to support new loads in `onResourceReady`: just a quick check to add in `GenericRequest.onResourceReady`, like: `listener != null && !listener.onResourceReady() && !isCancelled()`; probably need not call `notifyLoadSuccess` either.)

Adding `.skipMemoryCache(true)` to the inner load is another workaround, but the best solution is still `.thumbnail()` because it relies on no undocumented features, nor any timings.

By the way I found another reason why stealing the previous drawable is unsafe. If the ImageView is cleared the placeholder is re-set to it, further increasing the possibility of leaking it.

**Comment by sjudd** [comment]
This was "fixed" here: b06b0cc2bd79e10a77155466da8a83459edf371b. Right now it's an assertion. Dealing with the object pool makes handling the nested request gracefully a challenge. We could instead though add a new API that would take a request to start when the first one finishes, but I haven't seen that come up as a request.
